### ISSUE TITLE
Async + Streaming support for index/graph?

### ISSUE THREAD

**Issue by sirheimer** [jerryjliu/llama_index#927]
Hi,

I'm looking to implement streaming support in an async view. But there seems to be a bug at some level of the call that forgets to run, at the end in Langchain's BaseChatModel's `_generate` function. Seems like `_agenerate` function should be called. I've extensively gone through much of the calls and code for llama index to find where the issue originates, but with no luck.

It appears that the chain of aquery => QueryRunner => arun ... eventually leads to LLMPredictor.stream, which does not have a async equivalent.

This is the current code I'm running:

```python
from typing import Any, Dict, List, Optional, Type, Union
from asgiref.sync import async_to_sync
from langchain.callbacks.base import AsyncCallbackManager, AsyncCallbackHandler
from langchain.schema import LLMResult
from llama_index import LLMPredictor, PromptHelper
from llama_index.composability import ComposableGraph
from llama_index.indices.query.query_runner import QueryRunner
from llama_index.indices.query.query_transform.base import BaseQueryTransform
from llama_index.indices.query.schema import QueryBundle, QueryConfig
from llama_index.response.schema import Response
from langchain.chat_models import ChatOpenAI


class AsyncStreamingLLMCallbackHandler(AsyncCallbackHandler):
    async def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
        print(token)

    async def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:
        """Run when LLM ends running."""
        print("on_llm_end")

    async def on_llm_error(self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any) -> None:
        """Run when LLM errors."""
        print("on_llm_error")


QUERY_CONFIG_TYPE = Union[Dict, QueryConfig]


# The "use_async=True" parameter was missing so I built a custom ComposableGraph to add it, but did not help.
class AsyncComposableGraph(ComposableGraph):
    """Composable graph."""
    async def aquery(
        self,
        query_str: Union[str, QueryBundle],
        query_configs: Optional[List[QUERY_CONFIG_TYPE]] = None,
        llm_predictor: Optional[LLMPredictor] = None,
        query_transform: Optional[BaseQueryTransform] = None,
    ) -> Response:
        """Query the index."""
        # go over all the indices and create a registry
        llm_predictor = llm_predictor or self._llm_predictor
        query_runner = QueryRunner(
            llm_predictor,
            self._prompt_helper,
            self._embed_model,
            self._docstore,
            self._index_registry,
            query_configs=query_configs,
            query_transform=query_transform,
            recursive=True,
            use_async=True  # This parameter was missing so built custom ComposableGraph to add it, but did not help.
        )
        return await query_runner.aquery(query_str, self._index_struct)

llm = ChatOpenAI(
    streaming=True,
    callback_manager=AsyncCallbackManager([AsyncStreamingLLMCallbackHandler()]),
    verbose=True,
    temperature=0.4,
    presence_penalty=0.1,
    frequency_penalty=0.1,
    max_tokens=750,
    openai_api_key="sk...",
)

model = "gpt-4"

max_tokens=750
max_input_size = (7396 if model == "gpt-4" else 3796)
max_chunk_overlap = 20

prompt_helper = PromptHelper(max_input_size, max_tokens, max_chunk_overlap)

llm_predictor = LLMPredictor(llm=llm)

graph_path = './embeddings/indexed_graph.json'

graph = AsyncComposableGraph.load_from_disk(graph_path,llm_predictor=llm_predictor,prompt_helper=prompt_helper)

prompt = "Tell me about this information"

async def call():
    await graph.aquery(prompt)

async_to_sync(call)()

```

Here is the error information i get based on that code:

```shell
/usr/local/lib/python3.10/site-packages/langchain/chat_models/base.py:72: RuntimeWarning: coroutine 'AsyncCallbackManager.on_llm_start' was never awaited
  self.callback_manager.on_llm_start(
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

/usr/local/lib/python3.10/site-packages/langchain/chat_models/openai.py:256: RuntimeWarning: coroutine 'AsyncCallbackManager.on_llm_new_token' was never awaited
  self.callback_manager.on_llm_new_token(
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

/usr/local/lib/python3.10/site-packages/langchain/chat_models/base.py:80: RuntimeWarning: coroutine 'AsyncCallbackManager.on_llm_end' was never awaited
  self.callback_manager.on_llm_end(output, verbose=self.verbose)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

INFO 2023-03-27 16:26:14,082 token_counter 1 281472545759712 > [query] Total LLM token usage: 3581 tokens
INFO 2023-03-27 16:26:14,083 token_counter 1 281472545759712 > [query] Total embedding token usage: 5 tokens
```
It is only a warning and successfully generates an output, however the streaming functionality is not properly set up in async.

Would love to get this working and am happy to code the fixes myself, but am kind of stuck in the debugging of it. Thanks.

Eric

**Comment by sirheimer** [comment]
Also to note, the async streaming functionality works perfectly with a similar set up when not using the llama index and querying the LLM directly, and the streaming functionality itself works great using llama index without async. The combination is where the issue comes up.

Thanks in advance

**Comment by jerryjliu** [comment]
@sirheimer, we fixed streaming recently through this PR: https://github.com/jerryjliu/llama_index/pull/3125

Take a look at the streaming guide here: https://gpt-index.readthedocs.io/en/latest/how_to/customization/streaming.html

**Comment by manuel-84** [comment]
> @sirheimer, we fixed streaming recently through this PR: #3125
> 
> Take a look at the streaming guide here: https://gpt-index.readthedocs.io/en/latest/how_to/customization/streaming.html

cool.. it still doesn't include an example for streaming with `graph.as_query_engine` which seems doesnt include the parameter for streaming
